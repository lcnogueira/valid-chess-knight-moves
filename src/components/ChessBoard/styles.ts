import styled from 'styled-components'

export const ErrorMessage = styled.span`
  margin-bottom: 2rem;
  font-size: 1.4rem;
`

export const BoardGrid = styled.div`
  display: grid;
  grid-template-columns: repeat(8, 1fr);
  grid-template-rows: repeat(8, 1fr);
  width: 100%;
  max-width: 70rem;
  border: 1rem solid ${({ theme }) => theme.colors.white};
`
