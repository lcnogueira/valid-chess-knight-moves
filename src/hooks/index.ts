export { default as usePosition } from './usePosition'
