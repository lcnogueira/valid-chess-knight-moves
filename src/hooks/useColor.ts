import { useContext } from 'react'
import { ThemeContext } from 'styled-components'

function useTheme() {
  const themeContext = useContext(ThemeContext)

  return themeContext
}

export default useTheme
