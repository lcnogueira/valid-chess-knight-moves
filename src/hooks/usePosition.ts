import { useState, useEffect } from 'react'
import { getPossiblePositions } from 'services'

const usePosition = (initialPosition = '') => {
  const [currentPosition, setCurrentPosition] = useState(initialPosition)
  const [possiblePositions, setPossiblePositions] = useState<string[]>([])
  const [error, setError] = useState<string>('')

  const toggleCurrentPosition = (newPosition: string) => {
    //reset the error in case there is one
    if (error) setError('')
    //Unselect the current cell in case it's already selected or set the new position
    setCurrentPosition(currentPosition === newPosition ? '' : newPosition)
  }

  useEffect(() => {
    if (currentPosition) {
      getPossiblePositions(currentPosition)
        .then((positions) => setPossiblePositions(positions))
        .catch((error) => {
          setError(error.message)
          setPossiblePositions([])
        })
    } else {
      setPossiblePositions([])
    }
  }, [currentPosition])

  return { error, currentPosition, toggleCurrentPosition, possiblePositions }
}

export default usePosition
