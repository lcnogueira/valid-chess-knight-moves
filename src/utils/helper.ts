export const lines = ['8', '7', '6', '5', '4', '3', '2', '1']
export const columns = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

export const isValidNumber = (number: string) => lines.includes(number)

export const isValidLetter = (letter: string) =>
  columns.includes(letter.toUpperCase())

/**
 * The position is valid if componse by letter followed by one number and they both belong to the chess board limit.
 * @param position The position the user clicked on
 */
export const isAlgebraicNotation = (position: string) => {
  if (
    /^\D\d$/.test(position) &&
    isValidLetter(position.charAt(0)) &&
    isValidNumber(position.charAt(1))
  ) {
    return true
  }
  return false
}

/**
 * Returns the knight possible positions.
 * @param position The position the user clicked on
 */
export const findKnightPossibleMoves = (position: string): string[] => {
  if (position === 'A8') {
    return ['B6', 'C7']
  } else {
    return []
  }
}
