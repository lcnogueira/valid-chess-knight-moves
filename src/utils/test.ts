import {
  isAlgebraicNotation,
  isValidPosition,
  isValidIndex,
  findKnightPossibleMoves
} from './helper'

describe('helper methods', () => {
  test(`isAlgebraicNotation returns true if it's a valid notation`, () => {
    expect(isAlgebraicNotation('A1')).toBeTruthy()
    expect(isAlgebraicNotation('a1')).toBeTruthy()
    expect(isAlgebraicNotation('I1')).toBeTruthy()
    expect(isAlgebraicNotation('1A')).toBeFalsy()
    expect(isAlgebraicNotation('A11')).toBeFalsy()
  })

  test(`isValidPosition returns true if it's a valid notation and is in the chess board limit`, () => {
    expect(isValidPosition('A1')).toBeTruthy()
    expect(isValidPosition('A8')).toBeTruthy()
    expect(isValidPosition('h1')).toBeTruthy()
    expect(isValidPosition('H8')).toBeTruthy()
    expect(isValidPosition('A9')).toBeFalsy()
    expect(isValidPosition('H9')).toBeFalsy()
  })

  test('isValidIndex returns true if the index is equal or less than the array size', () => {
    expect(isValidIndex(0)).toBeTruthy()
    expect(isValidIndex(7)).toBeTruthy()
    expect(isValidIndex(-1)).toBeFalsy()
    expect(isValidIndex(8)).toBeFalsy()
  })

  test('findKnightPossibleMoves returns the posible positions where the knight can move', () => {
    expect(findKnightPossibleMoves('A8').sort()).toEqual(
      [
        'A4',
        'A6',
        'A8',
        'B5',
        'B6',
        'C4',
        'C7',
        'C8',
        'D5',
        'D7',
        'E6',
        'E8'
      ].sort()
    )
  })
})
