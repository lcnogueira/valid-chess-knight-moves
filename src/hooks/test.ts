import usePosition from './usePosition'
import { renderHook, act } from '@testing-library/react-hooks'
import fetchMock from 'fetch-mock'

describe('usePosition', () => {
  beforeAll(() => {
    global.fetch = window.fetch
  })
  afterAll(() => {
    fetchMock.restore()
  })

  it('Changes the possible positions according to changes of current position ', async () => {
    const position = 'A8'
    const possiblePositions = ['C7', 'B6']

    fetchMock.mock(`api/piece/knight/position/${position}?turns=2`, {
      status: 200,
      body: { positions: possiblePositions }
    })

    const { result } = renderHook(() => usePosition())

    //It starts with default values
    expect(result.current.currentPosition).toBe('')
    expect(result.current.possiblePositions).toEqual([])
    expect(result.current.error).toBe('')

    await act(async () => {
      result.current.toggleCurrentPosition(position)
    })

    //It changes the current position according to the toggleCurrentPosition
    expect(result.current.currentPosition).toBe(position)

    //The possible positions change accordingly to the change of currentPosition
    expect(result.current.possiblePositions).toEqual(possiblePositions)

    await act(async () => {
      result.current.toggleCurrentPosition(position)
    })

    //Unselects the position in case the user cliks on the same cell
    expect(result.current.currentPosition).toBe('')

    //Sets possible positions to an empty array if currentPosition is empty
    expect(result.current.possiblePositions).toEqual([])
  })

  it('Updates the error state with a message in case of an error', async () => {
    const wrongPosition = '1A'
    const message = 'You should provide a valid position'

    fetchMock.mock(`api/piece/knight/position/${wrongPosition}?turns=2`, {
      status: 400,
      body: { message, positions: [] }
    })

    const { result } = renderHook(() => usePosition())

    await act(async () => {
      result.current.toggleCurrentPosition(wrongPosition)
    })

    //If using a wrong position, the possible positions should be empty
    expect(result.current.possiblePositions).toEqual([])

    //It updates the error state with a message in case of an error
    expect(result.current.error).toBe(message)
  })
})
