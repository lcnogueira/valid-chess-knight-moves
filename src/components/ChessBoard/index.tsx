import { usePosition } from 'hooks'
import { lines, columns } from 'utils/helper'
import Cell from '../Cell'

import * as S from './styles'

const ChessBoard: React.FC = () => {
  const {
    error,
    currentPosition,
    toggleCurrentPosition,
    possiblePositions
  } = usePosition()

  const handleClick = (e: React.MouseEvent, cellPosition: string) => {
    e.preventDefault()
    toggleCurrentPosition(cellPosition)
  }

  return (
    <>
      {error && <S.ErrorMessage>{error}</S.ErrorMessage>}
      <S.BoardGrid>
        {lines.map((line, lineIndex) =>
          columns.map((column, columnIndex) => {
            const cellPosition = `${column}${line}`
            return (
              <Cell
                key={cellPosition}
                isOdd={(lineIndex + columnIndex) % 2 !== 0}
                onClick={(e) => handleClick(e, cellPosition)}
                isSelected={cellPosition === currentPosition}
                isPossiblePosition={possiblePositions.includes(cellPosition)}
              >
                {cellPosition}
              </Cell>
            )
          })
        )}
      </S.BoardGrid>
    </>
  )
}

export default ChessBoard
