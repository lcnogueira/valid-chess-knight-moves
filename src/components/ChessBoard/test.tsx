import { render, screen } from 'tests/utils'
import * as usePosition from 'hooks/usePosition'

import ChessBoard from '.'

describe('<ChessBoard />', () => {
  it('renders correctly', () => {
    const { container } = render(<ChessBoard />)

    expect(container).toMatchSnapshot()
  })

  it('shows a message error if it exists', () => {
    jest.spyOn(usePosition, 'default').mockReturnValue({
      error: 'An error ocurred',
      currentPosition: '',
      possiblePositions: [],
      toggleCurrentPosition: jest.fn()
    })

    render(<ChessBoard />)

    expect(screen.getByTestId('error')).toHaveTextContent('An error ocurred')
  })
})
