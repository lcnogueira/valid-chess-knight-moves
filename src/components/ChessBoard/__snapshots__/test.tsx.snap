// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChessBoard /> renders correctly 1`] = `
.c1 {
  background-color: #ffa500;
  height: 0;
  padding-top: 100%;
  border: none;
}

.c1:hover {
  opacity: 0.8;
}

.c2 {
  background-color: #FF6347;
  height: 0;
  padding-top: 100%;
  border: none;
}

.c2:hover {
  opacity: 0.8;
}

.c0 {
  display: grid;
  grid-template-columns: repeat(8,1fr);
  grid-template-rows: repeat(8,1fr);
  width: 100%;
  max-width: 70rem;
  border: 1rem solid #fff;
}

<div>
  
  <div
    class="c0"
  >
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
    <button
      class="c2"
    />
    <button
      class="c1"
    />
  </div>
</div>
`
